src/components/Sidebars/Payback.js

```diff
@@ -31,67 +31,93 @@ const Payback = ({ cdpId, reset }) => {
   const [daiBalance, setDaiBalance] = useState(0);
   const [liquidationPrice, setLiquidationPrice] = useState(0);
   const [collateralizationRatio, setCollateralizationRatio] = useState(0);
-  const [allowanceState, setAllowanceState] = useState({
-    proxyAddress: '',
-    userStartedWithoutAllowance: false,
+
+  const [userState, setUserState] = useState({
+    proxyAddress: null,
     allowance: null,
-    loading: false
+    startedWithoutProxy: false,
+    startedWithoutAllowance: false,
+    allowanceLoading: false,
+    proxyLoading: false
   });
 
   useEffect(() => {
-    const checkProxy = async () => {
-      const proxy = await maker.service('proxy').ensureProxy();
-      return proxy;
-    };
-
     const init = async () => {
-      const proxy = await checkProxy();
-      const allowance = await checkAllowance({ address: proxy });
-      setAllowanceState({
-        ...allowanceState,
-        proxyAddress: proxy,
-        userStartedWithoutAllowance: !allowance,
-        allowance
-      });
+      const proxyAddress = await checkProxy();
+      if (proxyAddress) {
+        const allowance = await checkAllowance(proxyAddress);
+        setUserState({
+          ...userState,
+          proxyAddress,
+          allowance,
+          startedWithoutAllowance: !allowance
+        });
+      } else {
+        setUserState({
+          ...userState,
+          startedWithoutProxy: true,
+          startedWithoutAllowance: true
+        });
+      }
     };
     init();
   }, []);
 
-  const checkAllowance = async ({ address }) => {
+  const checkAllowance = async address => {
     const daiToken = maker.getToken('MDAI');
     return (await daiToken.allowance(maker.currentAddress(), address)).eq(
       MAX_UINT_BN
     );
   };
 
+  const checkProxy = async () => {
+    const proxy = await maker.service('proxy').getProxyAddress();
+    return proxy;
+  };
+
+  const setProxy = async () => {
+    const { proxyAddress } = userState;
+    if (proxyAddress) return;
+
+    try {
+      setUserState({ ...userState, proxyLoading: true });
+      const txPromise = maker.service('proxy').ensureProxy();
+      newTxListener(txPromise, 'Setting Proxy Address');
+      await txPromise;
+
+      const proxyAddress = await checkProxy();
+      if (proxyAddress) {
+        setUserState({ ...userState, proxyAddress, proxyLoading: false });
+      } else {
+        setUserState({ ...userState, proxyLoading: false });
+      }
+    } catch (e) {
+      setUserState({ ...userState, proxyLoading: false });
+    }
+  };
+
   const setAllowance = async () => {
     // if allowance is already set, do nothing
-    if (allowanceState.allowance) return;
+    const { proxyAddress, allowance } = userState;
+    if (!proxyAddress) return;
```

Comment:
if there's already an allowance or proxy address is missing, can't we ensure this method is never called to begin with?

Reply:
I've replaced these conditional breaks to the `onClick` for both `setAllowance` and `setProxy`